pkg/server/backgroundprofiler/profiler/profiler.go

```diff
@@ -0,0 +1,342 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package profiler
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"runtime"
+	"runtime/pprof"
+	"runtime/trace"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/server/backgroundprofiler"
+	"github.com/cockroachdb/cockroach/pkg/server/dumpstore"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/pprofutil"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+	"github.com/google/pprof/profile"
+)
+
+var maxCombinedFileSize = settings.RegisterByteSizeSetting(
+	settings.TenantWritable,
+	"server.cpu_profiler.total_dump_size_limit",
+	"maximum combined disk size of preserved runtime traces",
+	128<<20, // 128MiB
+)
+
+const timeFormat = "2006-01-02T15_04_05.000"
+const CPUProfilerFileNamePrefix = "cpuprofiler."
+const RuntimeTraceFileNamePrefix = "runtimetrace."
+
+// BackgroundProfiler is a background service that runs on every node and is
+// capable of collecting on-demand CPU and execution profiles.
+//
+// The first Subscriber initializes the collection of the CPU and execution
+// trace profiles. While the profiles are being collected, only Subscribers with
+// the same `profileID` are allowed to join the running profiles. The profiles
+// are stopped and persisted to local storage when the last Subscriber
+// unsubscribes.
+type BackgroundProfiler struct {
+	nodeID int32
+
+	st        *cluster.Settings
+	stopper   *stop.Stopper
+	dumpStore *dumpstore.DumpStore
+
+	startProfile   chan struct{}
+	startedProfile chan struct{}
+	stopProfile    chan struct{}
+	res            chan protoutil.Message
+	mu             struct {
+		syncutil.Mutex
+
+		// profileID is a unique ID that can be associated with the operation that
+		// we are profiling. Only subscribers that are executing as part of this
```

Comment:
Is this too restrictive? If multiple traces are inflight across the cluster and touch on some node, one will fail, right? This may seem like an unimportant point now but traces need to be reliable. We already have two consumers of CPU profiler (status page/pprof endpoint, auto CPU profiler) and this new one is possibly more prolific.

It's a bit of a bigger lift (and an annoying one at that) but a pprof file is a protobuf and so is decomposable. We could allow multiple subscribers, have them share a single file, but associate the time span of each subscription in metadata along with the file (mapping subscriber ID to a slice of `[start,end)` timestamps). For that to work, the BackgroundProfiler should be in the path of any CPU profile.

This is all fairly onerous, I certainly don't think this should be in the first cut, but maybe it's worth pondering anyway.

Note that https://github.com/cockroachdb/cockroach/blob/4aa59bcc053a5ace5a0e9ea818b3d3d572b4eabe/pkg/settings/cluster/cluster_settings.go#L101-L114 is our current "mediator of access" to the profiling singleton, so here we're introducing a second one and should consolidate.